core/src/main/java/com/linecorp/armeria/internal/client/DefaultClientRequestContext.java

```diff
@@ -849,4 +850,33 @@ private String toStringSlow(@Nullable Channel ch, @Nullable RequestLogAccess par
             return buf.toString();
         }
     }
+
+    @Override
+    public CompletableFuture<Void> initiateConnectionShutdown() {
+        final CompletableFuture<Void> completableFuture = new CompletableFuture<>();
+
+        setAdditionalRequestHeader(HttpHeaderNames.CONNECTION, HttpHeaderValues.CLOSE);
+        log().whenRequestComplete().handle((ignore, ex) -> {
```

Comment:
There is a corner case where the connection is not closed forever even when `initiateConnectionShutdown()` is called.
- The response to the request is fully received
- keep-alive option is configured (NoopKeepAliveHandler will be used)
- No subsequent requests are sent to the session

In fact that I am not sure we need to consider the case if the idle timeout is disabled. However, it would be worth discussing.
We may define a new contract that this method should be called before a request headers is written to the transport layer.

What do you think of it?

Reply:
> We may define a new contract that this method should be called before a request headers is written to the transport layer.

I think such contract is not very intuitive. When a user calls `initiateConnectionShutdown()`, it's clear that a user expects the connection to be closed eventually, hopefully soon.

If the fix for this corner case is non-trivial, we could first merge this PR with a new known issue and then follow up there.

Reply:
> We may define a new contract that this method should be called before a request headers is written to the transport layer.

I agree with @trustin that it is more user-friendly to let the user call initiate shutdown regardless of the request timing. (It is a little confusing because a context is usually tied closely to a request though..)

I guess a fix for this will require some kind of scheduling which may be non-trivial, so +1 to handle separately